File: src/main/java/me/chell/samsara/api/gui/Drawable.java
package me.chell.samsara.api.gui;

import me.chell.samsara.api.util.WrapperJava;
import net.minecraft.client.gui.DrawableHelper;
import net.minecraft.client.util.math.MatrixStack;

public class Drawable implements WrapperJava {
    public int x, y, width, height;

    protected final MatrixStack matrices = new MatrixStack();

    public Drawable(int x, int y, int width, int height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public void draw(int mouseX, int mouseY) {
    }

    public boolean keyTyped(char typedChar, int keyCode) {
        return false;
    }

    public boolean mouseClicked(int mouseX, int mouseY, int mouseButton) {
        return false;
    }

    public void mouseReleased(int mouseX, int mouseY, int state) {
    }

    public void onGuiClosed() {
    }

    public void updateScreen() {
    }

    public void drawThemedString(String text, int x, int y) {
       getFontRenderer().drawWithShadow(matrices, text, x, y, GuiTheme.textColor.getValue().getArgb());
    }

    public void drawThemedStringRight(String text, int x, int y) {
        x -= getFontRenderer().getWidth(text);
        getFontRenderer().drawWithShadow(matrices, text, x, y, GuiTheme.textColor.getValue().getArgb());
    }

    public int getStringCenterX(int start, int end, String text) {
        return start + end/2 - getFontRenderer().getWidth(text)/2;
    }

    public int getStringCenterY(int start, int end) {
        return start + end/2 - getFontRenderer().fontHeight/2;
    }

    public void drawThemedRectPrimary(int x, int y, int width, int height) {
        drawRect(x, y, width, height, GuiTheme.primaryColor.getValue().getArgb());
    }

    public void drawThemedRectSecondary(int x, int y, int width, int height) {
        drawRect(x, y, width, height, GuiTheme.secondaryColor.getValue().getArgb());
    }

    public void drawThemedRectTertiary(int x, int y, int width, int height) {
        drawRect(x, y, width, height, GuiTheme.tertiaryColor.getValue().getArgb());
    }

    public void drawThemedBorder(int x, int y, int width, int height) {
        drawBorder(x, y, width, height, GuiTheme.borderColor.getValue().getArgb(), GuiTheme.borderSize.getValue());
    }

    public void drawBorder(int x, int y, int width, int height, int color, int size) {
        drawRect(x, y - size, width, size, color); // top
        drawRect(x, y + height, width, size, color); // bottom
        drawRect(x - size, y - size, size, height + size+size, color); // left
        drawRect(x + width, y - size, size, height + size+size, color); // right
    }

    public void drawRect(int x, int y, int width, int height, int color) {
        DrawableHelper.fill(matrices, x, y, x + width, y + height, color);
    }

    public void drawTexturedRect(int x, int y, int width, int height) {
        DrawableHelper.drawTexture(matrices, x, y, width, height, 0f, 0f, width, height, width, height);
    }
}

File: src/main/kotlin/me/chell/samsara/api/gui/Drawable.kt
package me.chell.samsara.api.gui

import me.chell.samsara.api.util.Align
import me.chell.samsara.api.util.Color
import me.chell.samsara.api.util.Globals
import net.minecraft.client.gui.DrawableHelper
import net.minecraft.client.util.math.MatrixStack

interface Drawable: Globals {

    fun render(matrices: MatrixStack, mouseX: Double, mouseY: Double, tickDelta: Float)
    fun tick()

    fun guiClosed()
    fun screenResized()
    fun themeLoaded()

    fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean
    fun keyReleased(keyCode: Int, scanCode: Int, modifiers: Int): Boolean
    fun charTyped(char: Char, modifiers: Int): Boolean

    fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean
    fun mouseReleased(mouseX: Double, mouseY: Double, button: Int): Boolean
    fun mouseScrolled(mouseX: Double, mouseY: Double, amount: Double): Boolean

    fun fill(matrices: MatrixStack, rect: Rectangle, color: Color) {
        DrawableHelper.fill(matrices, rect.x.toInt(), rect.y.toInt(), rect.endX.toInt(), rect.endY.toInt(), color.argb)
    }

    fun drawBorder(matrices: MatrixStack, border: Border, rect: Rectangle, color: Color) { // draw on the outside of rectangle
        if(border.top != 0.0) {
            val top = Rectangle(rect.x, rect.y - border.top, rect.width, border.top)
            fill(matrices, top, color)
        }
        if(border.bottom != 0.0) {
            val bottom = Rectangle(rect.x, rect.endY, rect.width, border.bottom)
            fill(matrices, bottom, color)
        }
        if(border.left != 0.0) {
            val left = Rectangle(rect.x - border.left, rect.y - border.top, border.left, rect.height + border.top + border.bottom)
            fill(matrices, left, color)
        }
        if(border.right != 0.0) {
            val right = Rectangle(rect.x + rect.width, rect.y - border.top, border.right, rect.height + border.top + border.bottom)
            fill(matrices, right, color)
        }
    }

    fun drawString(matrices: MatrixStack, text: String, properties: TextProperties, container: Rectangle) {
        val bounds = container.subtract(properties.padding)

        val x = when(properties.horizontal) {
            Align.Horizontal.LEFT -> bounds.x
            Align.Horizontal.RIGHT -> bounds.endX - fontRenderer.getWidth(text)
            Align.Horizontal.CENTER -> bounds.x + bounds.width / 2 - fontRenderer.getWidth(text) / 2
        }

        val y = when(properties.vertical) {
            Align.Vertical.TOP -> bounds.y
            Align.Vertical.BOTTOM-> bounds.endY - fontRenderer.fontHeight
            Align.Vertical.CENTER -> bounds.y + fontRenderer.fontHeight / 4
        }

        if(properties.shadow)
            fontRenderer.drawWithShadow(matrices, text, x.toFloat(), y.toFloat(), properties.color.argb)
        else
            fontRenderer.draw(matrices, text, x.toFloat(), y.toFloat(), properties.color.argb)
    }
}